File: 01-2.py
#!/usr/bin/env python

import re

lines = [row.strip() for row in open('01.input').readlines()]

numeral_pattern = re.compile("[1-9]|one|two|three|four|five|six|seven|eight|nine")

def name_to_numeral(name):
    numerals = {
        'one':   '1',
        'two':   '2',
        'three': '3',
        'four':  '4',
        'five':  '5',
        'six':   '6',
        'seven': '7',
        'eight': '8',
        'nine':  '9'
    }
    if name.isdigit():
        return name
    return numerals[name]


def scan_string(s):
    if s == '':
        return []
    elif m := re.match(numeral_pattern, s):
        return [m[0]] + scan_string(s[1:])
    else:
        return scan_string(s[1:])


sum = 0

for line in lines:
    numerals = scan_string(line)
    sum += int(name_to_numeral(numerals[0]) +
               name_to_numeral(numerals[-1]))

print(sum)

File: 01-2.rb
#!/usr/bin/env ruby

lines = File.read('01.input').lines.map(&:strip)

$numeral_pattern = /[[:digit:]]|one|two|three|four|five|six|seven|eight|nine/

def name_to_numeral(name)
  case name
  when /[[:digit:]]/
    name
  when 'one'
    '1'
  when 'two'
    '2'
  when 'three'
    '3'
  when 'four'
    '4'
  when 'five'
    '5'
  when 'six'
    '6'
  when 'seven'
    '7'
  when 'eight'
    '8'
  when 'nine'
    '9'
  end
end

def scan_string(s)
  if s.nil? or s.empty?
    []
  elsif s.start_with? $numeral_pattern
    [s.match($numeral_pattern).to_s] + scan_string(s[1..])
  else
    scan_string s[1..]
  end
end

print(lines.map do |line|
  numerals = scan_string line
  (name_to_numeral(numerals[0]) + name_to_numeral(numerals[-1])).to_i
end .sum, "\n")
